plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+)
+
+// serviceType is an OpenStack service type
+type serviceType string
+
+const (
+	// plugin is used to identify ourselves in log output
+	plugin                      = "openstack"
+	volumeV2Service serviceType = "volumev2"
+	Orchestration   serviceType = "orchestration"
+)
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint string          `toml:"authentication_endpoint"`
+	Domain           string          `toml:"domain"`
+	Project          string          `toml:"project"`
+	Username         string          `toml:"username"`
+	Password         string          `toml:"password"`
+	EnabledServices  []string        `toml:"enabled_services"`
+	ServerDiagnotics bool            `toml:"server_diagnotics"`
+	Timeout          config.Duration `toml:"timeout"`
+
+	// Locally cached clients
+	identity *gophercloud.ServiceClient
+	compute  *gophercloud.ServiceClient
+	volume   *gophercloud.ServiceClient
+	network  *gophercloud.ServiceClient
+	stack    *gophercloud.ServiceClient
+
+	// Locally cached resources
+	agents         map[string]agents.Agent
+	aggregates     map[int]aggregates.Aggregate
+	diag           map[string]interface{}
+	flavors        map[string]flavors.Flavor
+	hypervisors    map[string]hypervisors.Hypervisor
+	networks       map[string]networks.Network
+	nova_services  map[string]nova_services.Service
+	ports          map[string]ports.Port
+	projects       map[string]projects.Project
+	servers        map[string]servers.Server
+	services       map[string]services.Service
+	stacks         map[string]stacks.ListedStack
+	storagePools   map[string]schedulerstats.StoragePool
+	subnets        map[string]subnets.Subnet
+	volumes        map[string]volume
+	gather_servers bool
+
+	Log telegraf.Logger `toml:"-"`
+
+	tls.ClientConfig
+}
+
+// ContainsService indicates whether a particular service is enabled
+func (o *OpenStack) ContainsService(t serviceType) bool {
+	for _, service := range o.services {
+		if service.Type == string(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// Description returns a description string of the input plugin and implements
+// the Input interface.
+func (o *OpenStack) Description() string {
+	return "Collects performance metrics from OpenStack services"
+}
+
+// sampleConfig is a sample configuration file entry.
+var sampleConfig = `
+  ## The recommended interval to poll is '30m'
+
+  ## The identity endpoint to authenticate against and get the service catalog from.
+  authentication_endpoint = "https://my.openstack.cloud:5000"
+
+  ## The domain to authenticate against when using a V3 identity endpoint. Defaults to 'default'.
+  # domain = "default"
+
+  ## The project to authenticate as. Defaults to 'admin'.
+  # project = "admin"
```

Comment:
Please remove the `Defaults to` comments as it is redundant with the declaration below.